crates/runtime/src/accelerated_table/refresh_task/changes.rs

```diff
@@ -0,0 +1,239 @@
+                    let primary_keys = change_batch.primary_keys(row);
+                    let ctx = SessionContext::new();
+                    let session_state = ctx.state();
+
+                    tracing::info!(
+                        "Upserting data row for {dataset_name} with {}",
+                        Self::get_primary_key_log_fmt(&inner_data, &primary_keys)?
+                    );
+
+                    let insert_plan = self
+                        .accelerator
+                        .insert_into(
+                            &session_state,
+                            Arc::new(DataUpdateExecutionPlan::new(DataUpdate {
+                                schema: inner_data.schema(),
+                                data: vec![inner_data],
+                                update_type: UpdateType::Append,
+                            })),
+                            false,
+                        )
+                        .await
+                        .context(crate::accelerated_table::FailedToWriteDataSnafu)?;
+
+                    collect(insert_plan, ctx.task_ctx())
+                        .await
+                        .context(crate::accelerated_table::FailedToWriteDataSnafu)?;
+                }
+                _ => {
+                    tracing::error!("Unknown operation {op} for {dataset_name}");
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_primary_key_log_fmt(
+        data: &RecordBatch,
+        primary_keys: &[String],
+    ) -> crate::accelerated_table::Result<String> {
+        primary_keys
+            .iter()
+            .map(|key| {
+                let (value, _) = Self::get_primary_key_value(data, key)?;
+                Ok(format!("{key}={value}"))
+            })
+            .collect::<crate::accelerated_table::Result<Vec<String>>>()
+            .map(|keys| keys.join(", "))
+    }
+
+    fn get_delete_where_expr(
+        data: &RecordBatch,
+        primary_keys: Vec<String>,
+    ) -> crate::accelerated_table::Result<Vec<Expr>> {
+        let mut delete_where_exprs: Vec<Expr> = vec![];
+
+        for primary_key in primary_keys {
+            let (_, expr_val) = Self::get_primary_key_value(data, &primary_key)?;
+            delete_where_exprs.push(col(primary_key).eq(expr_val));
+        }
+
+        Ok(delete_where_exprs)
+    }
+
+    fn get_primary_key_value(
+        data: &RecordBatch,
+        key: &str,
+    ) -> crate::accelerated_table::Result<(String, Expr)> {
+        let data_schema = data.schema();
+        let (primary_key_idx, field) = data_schema.column_with_name(key).ok_or_else(|| {
+            crate::accelerated_table::ExpectedSchemaToHaveFieldSnafu {
+                field_name: key.to_string(),
+                schema: Arc::clone(&data_schema),
+            }
+            .build()
+        })?;
+
+        let key_col = data.column(primary_key_idx);
+        match field.data_type() {
```

Comment:
Remind me why primary keys in this context are limited to just `{Int32, Int64, Utf8}`?

Reply:
Okay, its a limitation (see `PrimaryKeyTypeNotYetSupportedSnafu `).

Reply:
It covers the most common types. We can get feedback on additional types we need to support once its out.